src/entry.c

```diff
@@ -72,74 +82,85 @@ sds entryGetField(const entry *entry) {
 
 /* Returns the location of a pointer to a separately allocated value. Only for
  * an entry without an embedded value. */
-static sds *entryGetValueRef(const entry *entry) {
+static inline void **entryGetValueRef(const entry *entry) {
     serverAssert(entryHasValuePtr(entry));
     char *field_data = sdsAllocPtr(entry);
-    field_data -= sizeof(sds);
-    return (sds *)field_data;
+    field_data -= sizeof(void *);
+    return (void **)field_data;
 }
 
-/* Returns the sds of the entry's value. */
-sds entryGetValue(const entry *entry) {
-    if (entryHasValuePtr(entry)) {
-        return *entryGetValueRef(entry);
-    } else {
+/* Returns the entry's value. */
+char *entryGetValue(const entry *entry, size_t *len) {
+    if (entryHasEmbeddedValue(entry)) {
         /* Skip field content, field null terminator and value sds8 hdr. */
         size_t offset = sdslen(entry) + 1 + sdsHdrSize(SDS_TYPE_8);
-        return (char *)entry + offset;
+        sds value = (char *)entry + offset;
+        if (len) *len = sdslen(value);
+        return value;
     }
+    void **value_ref = entryGetValueRef(entry);
+    if (!value_ref) return NULL;
+    if (entryHasStringRef(entry)) {
+        stringRef *string_ref = (stringRef *)*value_ref;
+        if (len) *len = string_ref->len;
+        return (char *)string_ref->buf;
+    }
+    if (len) *len = sdslen(*value_ref);
+    return *value_ref;
 }
 
+static void entrySetValuePtr(entry *e, sds value) {
```

Comment:
Lets be more explicit
```suggestion
static void entrySetValueSds(entry *e, sds value) {
```